### PR TITLE
Add chart for backup-restore-operator

### PR DESCRIPTION
**Pull request by mrajashree**
This chart needs to create a CR when its installed. So as per offline discussion of making original charts similar in structure to upstream chart, this chart has all its CRDs in the /crds folder. And the CR it should create is in the templates folder.
Carrying over all relevant comments from https://github.com/rancher/charts/pull/589

This chart creates:
1. ServiceAccount bound to the cluster-admin role, which is used for launching the operator deployment. This operator needs CRUD permissions on all resources in the cluster. Hence only cluster-admin should be able to operate it. That's why it was decided in design meetings to assign it the cluster-admin role and have no other roles for it.
2. The CRDs in crds folder are generated using wrangler's CRD definition and openapi schema package. [This](https://rancher.atlassian.net/wiki/spaces/EN/pages/494469147/Backup+Restore+Rancher+MCM+app) design doc covers the explanation behind each field
3. User can either configure a s3 bucket OR a PVC to store backups in. The operator will error if both are specified. This helm chart also uses the `fail` template function to prevent chart installation if both are enabled.
4. If s3.enabled is set to true, the chart creates a secret in the chart's namespace that contains all details required for uploading/retrieving backups to/from s3 compatible stores.
5. If persistence.enabled is set to true, the operator creates a PVC in the chart's namespace and mounts it at /var/lib/backups. We want the PVC's access mode to be RWO, hence that is not accepted from the user. The user can specify storageClass, volumeName and size. The fields accepted for this PVC and the comments surrounding them in the values.yaml file follow the pattern that most of the upstream Helm charts follow. 
6. The `checksum/secret` annotation for s3-secret in the deployment spec will re-configure the pod to use the updated s3 secret if the user changes their s3 configuration, fixing https://github.com/rancher/rancher/issues/28607

https://github.com/rancher/rancher/issues/28640

NOTE: The "url" field in package.yaml is empty since its an original chart, will change it based on the new logic for generating CRDs.

https://github.com/rancher/rancher/issues/28368

### CODE REVIEW

File: packages/backup-restore-operator/charts/values.yaml
Comment (StrongMonkey) [inline packages/backup-restore-operator/charts/values.yaml:15]: Do you need to comment this out `(optional)`?

File: packages/backup-restore-operator/charts/values.yaml
Comment (StrongMonkey) [inline packages/backup-restore-operator/charts/values.yaml:17]: same above

File: packages/backup-restore-operator/package.yaml
Comment (StrongMonkey) [inline packages/backup-restore-operator/package.yaml:1]: Yeah you need to fix CI so that this doesn't error out.

File: packages/backup-restore-operator/package.yaml
Comment (StrongMonkey) [inline packages/backup-restore-operator/package.yaml:2]: Also don't have to set packageVersion in here

File: packages/backup-restore-operator/package.yaml
Comment (mrajashree) [reply packages/backup-restore-operator/package.yaml:1]: yup, I don't think it's decided yet what the expected value should be?

Comment (mrajashree) [reply packages/backup-restore-operator/package.yaml:1]: yes, I don't think its decided yet what the value for `url` should be for the scripts in case of original chart

Comment (mrajashree) [reply packages/backup-restore-operator/package.yaml:1]: yes, it's being tracked in this issue https://github.com/rancher/rancher/issues/28628

File: packages/backup-restore-operator/charts/templates/serviceaccount.yaml
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/templates/serviceaccount.yaml:7]: nit, `nindent` should automatically indent the first line iirc. Happens a couple of times throughout the PR.

```suggestion
  labels: {{- include "backupRestore.labels" . | nindent 4 }}
```

File: packages/backup-restore-operator/charts/templates/s3-secret.yaml
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/templates/s3-secret.yaml:11]: nit:
```suggestion
  {{- with .Values.s3 }}
```

File: packages/backup-restore-operator/charts/templates/s3-secret.yaml
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/templates/s3-secret.yaml:28]: Should this be set to a default value or does the fact that it doesn't exist imply that it's false?

i.e.
```suggestion
  insecureTLSSkipVerify: {{ default false .insecureTLSSkipVerify }} 
```

File: packages/backup-restore-operator/charts/templates/deployment.yaml
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/templates/deployment.yaml:41]: nit:
```suggestion
          {{- end }}
```

File: packages/backup-restore-operator/charts/templates/s3-secret.yaml
Comment (mrajashree) [reply packages/backup-restore-operator/charts/templates/s3-secret.yaml:28]: yup the fact that it doesn't exist implies that it's false

File: packages/backup-restore-operator/charts/templates/serviceaccount.yaml
Comment (mrajashree) [reply packages/backup-restore-operator/charts/templates/serviceaccount.yaml:7]: yeah I was using the sample chart helm3 creates and it added the labels in this format (next line) so thought I'd let it be 😅

File: packages/backup-restore-operator/charts/templates/_helpers.tpl
Comment (cbron) [inline packages/backup-restore-operator/charts/templates/_helpers.tpl:67]: I know this was in example, but by 2.5.0 we will be shipping 1.16+, so that greatly simplifies it. We could break out of helper, up to you.

File: packages/backup-restore-operator/charts/templates/deployment.yaml
Comment (cbron) [inline packages/backup-restore-operator/charts/templates/deployment.yaml:52]: I thought you could only do one nodeSelector in a pod spec. Are you trying to allow this to go on either type of node ?

File: packages/backup-restore-operator/package.yaml
Comment (cbron) [inline packages/backup-restore-operator/package.yaml:3]: Recent development, can you put this line up a level where url was.

**Review by cbron — CHANGES_REQUESTED** [review]
CI will be fixed once Brenda's PR is in, we should wait to merge until then or it will break all other chart generation.

File: packages/backup-restore-operator/charts/templates/deployment.yaml
Comment (mrajashree) [reply packages/backup-restore-operator/charts/templates/deployment.yaml]: yeah even with both nodeSelectors the chart was able to run the pod on a linux cluster node. Although after it was created the pod only had the `kubernetes.io/os: linux` nodeSelector. So looks like with both selectors specified it allows the pod to go on either type of node. But I haven't tested it for a windows node, so I removed that selector

Comment (cbron) [reply packages/backup-restore-operator/charts/templates/deployment.yaml]: I think it just overwrites with whatever was set last. If you want to test you could flip them and try to deploy to linux and see if it fails.

Comment (cbron) [reply packages/backup-restore-operator/charts/templates/deployment.yaml]: Does this need to run on windows for 2.5.0 ? Might be able to solve this with nodeAffinity, we can talk about it.

**Review by cbron — COMMENTED** [review]
Are we not doing the User RBAC roles for admin/view/edit ? Not sure if it makes sense for your use case. Outside of that and windows node selector it lgtm.

**Comment by mrajashree** [comment]
@cbron yes individual roles for admin/edit/view won't be applicable, because each restore operation can modify/delete any resources. I have removed the windows node selector based on our discussion

File: packages/backup-restore-operator/charts/templates/deployment.yaml
Comment (mrajashree) [reply packages/backup-restore-operator/charts/templates/deployment.yaml]: based on offline discussion, we don't need to add it in for windows right now

**Review by cbron — APPROVED** [review]
lgtm, just wait for other PR to merge, rebase to re-run CI to make sure it passes

File: packages/backup-restore-operator/charts/values.yaml
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/values.yaml:49]: nit:
```suggestion
    ## Path that is exported by the NFS server. More info: https://kubernetes.io/docs/concepts/storage/volumes#nfs
    path: ""
    
    ## Server is the hostname or IP address of the NFS server. More info: https://kubernetes.io/docs/concepts/storage/volumes#nfs
    server: ""
```

File: packages/backup-restore-operator/charts/templates/nfs-pv.yaml
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/templates/nfs-pv.yaml:18]: nit: newline

File: packages/backup-restore-operator/charts/templates/deployment.yaml
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/templates/deployment.yaml:60]: nit: the unindented include looks weird. We should either indent it or move it back to the previous line like:
```suggestion
      tolerations: {{- include "linux-node-tolerations" . | nindent 8}}
```

File: packages/backup-restore-operator/charts/templates/_helpers.tpl
Comment (aiyengar2) [inline packages/backup-restore-operator/charts/templates/_helpers.tpl:70]: nit: inconsistent casing

`linux-node-tolerations` uses a different format than the other helpers in this file. Should it be `backupRestore.linuxNode.tolerations`? Would also be nice it we do something similar with `kubernetes.io/os: linux` for `backupRestore.linuxNode.selector`.

Alternatively, since this is something Rancher-specific but not backup-restore specific, maybe we should add it to the top of the file under `system_default_registry` as `linux_node_tolerations`.

**Review by aiyengar2 — APPROVED** [review]
nit comments, LGTM

**Review by StrongMonkey — APPROVED** [review]

File: packages/backup-restore-operator/charts/templates/_helpers.tpl
Comment (mrajashree) [reply packages/backup-restore-operator/charts/templates/_helpers.tpl:70]: the node selector was just one line, and from a previous comment I thought we could break out of the helper for that. I can rename the tolerations helper though

File: packages/backup-restore-operator/charts/values.yaml
Comment (mrajashree) [reply packages/backup-restore-operator/charts/values.yaml]: not applicable anymore, after some discussion we decided to remove nfs part

**Pull request merged**
